File: sensor-monitor/python/sensor_monitor/sliding_window.py
from sensor_monitor.ring_buffer import RingBuffer
class SlidingWindow:
    """Sliding window statistics over a fixed-size ring buffer."""

    def __init__(self, window_size: int):
        self._buffer = RingBuffer(window_size)

    def push(self, value: float):
        """Add a reading to the window."""
        self._buffer.push(value)

    def average(self) -> float:
        """Compute the mean of the current window."""
        if len(self._buffer) == 0:
            return 0.0
        total = sum(self._buffer)
        return total / len(self._buffer)

    def maximum(self) -> float:
        """Return the max of the current window."""
        if len(self._buffer) == 0:
            return float("-inf")
        return max(self._buffer)

    def minimum(self) -> float:
        """Return the min of the current window."""
        if len(self._buffer) == 0:
            return float("inf")
        return min(self._buffer)

    def __len__(self) -> int:
        return len(self._buffer)

File: sensor-monitor/rust/src/sliding_window.rs
use crate::ring_buffer::RingBuffer;

/// Sliding window statistics over a fixed-size ring buffer.
pub struct SlidingWindow {
    buffer: RingBuffer<f64>,
}

impl SlidingWindow {
    pub fn new(window_size: usize) -> Self {
        SlidingWindow {
            buffer: RingBuffer::new(window_size),
        }
    }

    pub fn push(&mut self, value: f64) {
        self.buffer.push(value);
    }

    pub fn average(&self) -> f64 {
        if self.buffer.len() == 0 {
            return 0.0;
        }
        let sum: f64 = self.buffer.iter().sum();
        sum / self.buffer.len() as f64
    }

    pub fn maximum(&self) -> f64 {
        self.buffer
            .iter()
            .copied()
            .fold(f64::NEG_INFINITY, f64::max)
    }

    pub fn minimum(&self) -> f64 {
        self.buffer
            .iter()
            .copied()
            .fold(f64::INFINITY, f64::min)
    }

    pub fn len(&self) -> usize {
        self.buffer.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sliding_window() {
        let mut sw = SlidingWindow::new(3);
        sw.push(10.0); sw.push(20.0); sw.push(30.0);
        assert!((sw.average() - 20.0).abs() < 1e-10);
        assert!((sw.maximum() - 30.0).abs() < 1e-10);
        assert!((sw.minimum() - 10.0).abs() < 1e-10);
        sw.push(40.0);
        assert!((sw.average() - 30.0).abs() < 1e-10);
        assert!((sw.maximum() - 40.0).abs() < 1e-10);
    }
}